url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/DNSMessage.java

```diff
@@ -0,0 +1,1133 @@
+        questions = new ArrayList<>(questionCount);
+        for (int i = 0; i < questionCount; i++) {
+            questions.add(new Question(dis, data));
+        }
+        answerSection = new ArrayList<>(answerCount);
+        for (int i = 0; i < answerCount; i++) {
+            answerSection.add(Record.parse(dis, data));
+        }
+        authoritySection = new ArrayList<>(nameserverCount);
+        for (int i = 0; i < nameserverCount; i++) {
+            authoritySection.add(Record.parse(dis, data));
+        }
+        additionalSection = new ArrayList<>(additionalResourceRecordCount);
+        for (int i = 0; i < additionalResourceRecordCount; i++) {
+            additionalSection.add(Record.parse(dis, data));
+        }
+        optRrPosition = getOptRrPosition(additionalSection);
+    }
+
+    /**
+     * Constructs an normalized version of the given DNSMessage by setting the id to '0'.
+     *
+     * @param message the message of which normalized version should be constructed.
+     */
+    private DNSMessage(DNSMessage message) {
+        id = 0;
+        qr = message.qr;
+        opcode = message.opcode;
+        authoritativeAnswer = message.authoritativeAnswer;
+        truncated = message.truncated;
+        recursionDesired = message.recursionDesired;
+        recursionAvailable = message.recursionAvailable;
+        authenticData = message.authenticData;
+        checkingDisabled = message.checkingDisabled;
+        responseCode = message.responseCode;
+        receiveTimestamp = message.receiveTimestamp;
+        questions = message.questions;
+        answerSection = message.answerSection;
+        authoritySection = message.authoritySection;
+        additionalSection = message.additionalSection;
+        optRrPosition = message.optRrPosition;
+    }
+
+    private static int getOptRrPosition(List<Record<? extends Data>> additionalSection) {
+        int optRrPosition = -1;
+        for (int i = 0; i < additionalSection.size(); i++) {
+            Record<? extends Data> record = additionalSection.get(i);
+            if (record.type == Record.TYPE.OPT) {
+                optRrPosition = i;
+                break;
+            }
+        }
+        return optRrPosition;
+    }
+
+    /**
+     * Generate a binary dns packet out of this message.
+     *
+     * @return byte[] the binary representation.
+     * @throws IOException Should never happen.
+     */
+    public byte[] toArray() throws IOException {
+        return serialize().clone();
+    }
+
+    public DatagramPacket asDatagram(InetAddress address, int port) {
+        byte[] bytes = serialize();
+        return new DatagramPacket(bytes, bytes.length, address, port);
+    }
+
+    public void writeTo(DataOutputStream dataOutputStream) throws IOException {
+        byte[] bytes = serialize();
+        dataOutputStream.writeShort(bytes.length);
+        dataOutputStream.write(bytes);
+    }
+
+    private byte[] byteCache;
+
+    private byte[] serialize() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.